src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.SyncNamespaceCodeAction.cs

```diff
@@ -0,0 +1,322 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;         
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceCodeRefactoringProvider<TService, TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TService : AbstractSyncNamespaceCodeRefactoringProvider<TService, TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode 
+    {
+
+        private enum ActionKind
+        {
+            MoveFile,
+            ChangeNamespace,
+        }
+
+        private class SyncNamespaceCodeAction : CodeAction
+        {
+            public override string Title
+            {
+                get
+                {
+                    switch (_kind)
+                    {
+                        case ActionKind.ChangeNamespace:
+                            return "Change namespace!";
+                        case ActionKind.MoveFile:
+                            return "Move file!";
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(_kind);
+                    }
+                }
+            }
+
+            private readonly ActionKind _kind;
```

Comment:
kind flag seems to serve strange purpose.  instead of having one code-action, with kinds to determine behavior... why not just have different code actions?